File: rewrite-go/pkg/parser/modgraph/tidy.go
/*
 * Copyright 2026 the original author or authors.
 *
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://docs.moderne.io/licensing/moderne-source-available-license
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package modgraph

import (
	"strings"

	"golang.org/x/mod/modfile"
	"golang.org/x/mod/module"
	"golang.org/x/mod/semver"
)

// TidyRequireSet computes the exact go.mod require set `go mod tidy` would write,
// including the go 1.17+ pruning-completeness roots that NeededModules alone does
// not capture.
//
// NeededModules gives the modules that PROVIDE a package in `all` (direct +
// import-reachable indirect). For a go>=1.17 main module, `go mod tidy` also adds
// an indirect root for every module that is reachable from `all` through imports
// OR TESTS but whose version the pruned module graph would UNDER-SELECT — i.e. a
// test-transitive dependency whose required version is higher than any root's
// pruned go.mod implies (e.g. kr/text via gopkg.in/check.v1, go.uber.org/mock via
// a dependency's test). This mirrors cmd/go/internal/modload.tidyPrunedRoots:
// walk imports+tests outward from `all`, and promote a module to an explicit root
// whenever Selected(path) < loadedVersion(path).
//
// The version gate is essential: it adds mock/kr/text (genuinely under-selected)
// while leaving testify-style clusters out when the pruned graph already selects
// them correctly. For go<1.17 (no pruning) or an incomplete base resolution it
// returns NeededModules unchanged.
func TidyRequireSet(mainImports []string, mainModulePath, mainGoMod string, res Result, src ModSource, separateIndirect bool) RequireSet {
	base := NeededModules(mainImports, mainModulePath, res, src, separateIndirect)
	if !separateIndirect || !base.Complete {
		return base
	}

	// loaded[path] = the version each module is selected at in the full build
	// list (go's `m.Version` — the version a package's module is loaded at).
	loaded := map[string]string{}
	var mods []modVer
	for _, m := range res.BuildList {
		if !m.Main {
			loaded[m.Path] = m.Version
			mods = append(mods, modVer{m.Path, m.Version})
		}
	}
	moduleOf := func(importPath string) (string, string) {
		best, bestVer := "", ""
		for _, m := range mods {
			if importPath == m.path || strings.HasPrefix(importPath, m.path+"/") {
				if len(m.path) > len(best) {
					best, bestVer = m.path, m.version
				}
			}
		}
		return best, bestVer
	}
	isLocal := func(importPath string) bool {
		return importPath == mainModulePath || strings.HasPrefix(importPath, mainModulePath+"/")
	}

	// Current root set: everything NeededModules already requires.
	roots := map[string]string{}
	for p, v := range base.Direct {
		roots[p] = v
	}
	for p, v := range base.Indirect {
		roots[p] = v
	}

	// Build a requirement index once: every module version's direct requires,
	// seeded for free from the already-resolved graph and lazily fetching only
	// the few promoted roots that pruning left unloaded. Pruned MVS then runs
	// entirely in memory — no per-iteration re-resolution.
	idx := newReqIndex(res, mainGoMod, src)

	// Walk the package import graph FRONTIER BY FRONTIER, by increasing import-
	// stack depth, mirroring cmd/go/internal/modload.tidyPrunedRoots. At each
	// frontier we recompute the pruned selection under the roots accumulated so
	// far, then promote any frontier module the pruned graph under-selects. This
	// ordering is essential: promoting a shallow module (e.g. kr/pretty) pins its
	// deeper requirements (kr/text) BEFORE they are examined, so they are not
	// wrongly promoted. A package's TEST imports are deferred one frontier deeper
	// (go models this as a separate `<pkg>.test` node) so test-transitive deps
	// sort below ordinary ones.
	type qitem struct {
		path   string
		isTest bool
	}
	queued := map[string]bool{}
	var queue []qitem
	enq := func(path string, isTest bool) {
		if isStdlibImport(path) || isLocal(path) {
			return
		}
		k := path
		if isTest {
			k += "\x00t"
		}
		if !queued[k] {
			queued[k] = true
			queue = append(queue, qitem{path, isTest})
		}
	}
	for _, imp := range mainImports {
		enq(imp, false)
	}

	for len(queue) > 0 {
		sel := prunedSelectInMemory(roots, idx)
		frontier := queue
		queue = nil
		for _, it := range frontier {
			mod, ver := moduleOf(it.path)
			if mod == "" {
				continue
			}
			imports, testImports, err := packageImportsWithTests(src, mod, ver, it.path)
			if err == nil {
				if it.isTest {
					for _, d := range testImports {
						enq(d, false)
					}
				} else {
					for _, d := range imports {
						enq(d, false)
					}
					enq(it.path, true) // the package's test node, one frontier deeper
				}
			}
			if _, isRoot := roots[mod]; !isRoot {
				want := loaded[mod]
				have := sel[mod]
				if want != "" && (have == "" || semver.Compare(have, want) < 0) {
					roots[mod] = want
				}
			}
		}
	}

	// Reclassify: direct stays direct; every other root is indirect.
	out := RequireSet{
		Direct:      map[string]string{},
		Indirect:    map[string]string{},
		Complete:    true,
		Unresolved:  base.Unresolved,
		MissingDirs: base.MissingDirs,
	}
	for p, v := range roots {
		if _, isDirect := base.Direct[p]; isDirect {
			out.Direct[p] = v
		} else {
			out.Indirect[p] = v
		}
	}
	return out
}

// reqIndex memoizes each module version's direct requirements (post-replace) and
// go directive. It is seeded for free from an already-resolved graph (whose edges
// cover every LOADED module) and lazily fetches the go.mod of any module that
// pruning left unloaded — only ever the handful of pruned modules promoted to
// roots. This lets prunedSelectInMemory run a pruned MVS without re-resolving.
type reqIndex struct {
	edges   map[string][]module.Version // "path@version" -> requires (post-replace)
	gover   map[string]string           // "path@version" -> go directive
	known   map[string]bool             // keys whose edges are fully populated
	replace map[string]module.Version   // main module's version replacements
	src     ModSource
}

func newReqIndex(res Result, mainGoMod string, src ModSource) *reqIndex {
	idx := &reqIndex{
		edges:   map[string][]module.Version{},
		gover:   map[string]string{},
		known:   map[string]bool{},
		replace: map[string]module.Version{},
		src:     src,
	}
	// Main module's version replacements (local-path replaces are skipped; they
	// already mark the resolution incomplete upstream).
	if mf, err := modfile.Parse("go.mod", []byte(mainGoMod), nil); err == nil {
		for _, r := range mf.Replace {
			if r.New.Version == "" {
				continue
			}
			idx.replace[r.Old.Path] = r.New
			idx.replace[r.Old.Path+"@"+r.Old.Version] = r.New
		}
	}
	// Seed edges from the resolved graph. Every module that was LOADED contributes
	// all of its require edges here, so its key is fully known.
	for _, e := range res.Graph {
		key := e.FromPath + "@" + e.FromVersion
		idx.edges[key] = append(idx.edges[key], module.Version{Path: e.ToPath, Version: e.ToVersion})
		idx.known[key] = true
	}
	for _, m := range res.BuildList {
		idx.gover[m.Path+"@"+m.Version] = m.GoVersion
	}
	return idx
}

func (idx *reqIndex) applyReplace(m module.Version) module.Version {
	if nv, ok := idx.replace[m.Path+"@"+m.Version]; ok {
		return nv
	}
	if nv, ok := idx.replace[m.Path]; ok {
		return nv
	}
	return m
}

// requires returns module path@version's direct requirements and go directive,
// fetching and memoizing the go.mod when not already seeded.
func (idx *reqIndex) requires(path, version string) ([]module.Version, string) {
	key := path + "@" + version
	if idx.known[key] {
		return idx.edges[key], idx.gover[key]
	}
	idx.known[key] = true // memoize even on miss, so a failed fetch isn't retried
	b, ok := idx.src.GoMod(path, version)
	if !ok {
		return nil, idx.gover[key]
	}
	df, err := modfile.Parse(key, b, nil)
	if err != nil {
		return nil, idx.gover[key]
	}
	if df.Go != nil {
		idx.gover[key] = df.Go.Version
	}
	reqs := make([]module.Version, 0, len(df.Require))
	for _, r := range df.Require {
		reqs = append(reqs, idx.applyReplace(r.Mod))
	}
	idx.edges[key] = reqs
	return reqs, idx.gover[key]
}

// prunedSelectInMemory computes the MVS-selected version of every module under the
// go>=1.17 PRUNED module graph rooted at the given root set, reading requirements
// from idx. It mirrors Resolve's traversal exactly: every loaded module's requires
// become build-list nodes, but a module's requires are recursed into only when the
// module is unpruned (its go directive is < 1.17).
func prunedSelectInMemory(roots map[string]string, idx *reqIndex) map[string]string {
	present := map[string]string{} // path -> selected version
	loadPath := map[string]bool{}  // paths we recurse into
	enqueued := map[string]bool{}
	type pv struct{ path, version string }
	var queue []pv

	enqueue := func(m module.Version) {
		k := m.Path + "@" + m.Version
		if !enqueued[k] {
			enqueued[k] = true
			queue = append(queue, pv{m.Path, m.Version})
		}
	}
	setNode := func(m module.Version) {
		if v, ok := present[m.Path]; !ok || semver.Compare(m.Version, v) > 0 {
			present[m.Path] = m.Version
			if loadPath[m.Path] {
				enqueue(m)
			}
		}
	}
	markLoad := func(m module.Version) {
		loadPath[m.Path] = true
		enqueue(m)
	}

	// Roots = the synthetic main module's requirements.
	for path, ver := range roots {
		m := module.Version{Path: path, Version: ver}
		setNode(m)
		markLoad(m)
	}

	for len(queue) > 0 {
		cur := queue[0]
		queue = queue[1:]
		if present[cur.path] != cur.version {
			continue // superseded by a higher selected version
		}
		reqs, goV := idx.requires(cur.path, cur.version)
		unpruned := goUnpruned(goV)
		for _, req := range reqs {
			setNode(req)
			if unpruned {
				markLoad(req)
			}
		}
	}
	return present
}

// modVer is a build-list module path at its selected version.
type modVer struct{ path, version string }

File: rewrite-go/src/main/java/org/openrewrite/golang/internal/modgraph/Tidy.java
/*
 * Copyright 2026 the original author or authors.
 * <p>
 * Licensed under the Moderne Source Available License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://docs.moderne.io/licensing/moderne-source-available-license
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.golang.internal.modgraph;

import org.jspecify.annotations.Nullable;

import java.nio.charset.StandardCharsets;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Computes the exact go.mod require set {@code go mod tidy} would write, by
 * walking the PACKAGE import graph (not just the module graph). {@link
 * #neededModules} gives the modules that provide a package in {@code all};
 * {@link #tidyRequireSet} additionally adds the go1.17+ pruning-completeness
 * roots — test-transitive modules the pruned graph would under-select — mirroring
 * {@code cmd/go/internal/modload.tidyPrunedRoots}. No {@code go} execution.
 */
public final class Tidy {

    private Tidy() {
    }

    public static RequireSet neededModules(List<String> mainImports, String mainModulePath,
                                           ResolveResult res, ModSource src, boolean separateIndirect) {
        RequireSet rs = new RequireSet();
        List<ResolveResult.Module> mods = nonMain(res);

        Map<String, String> needed = new HashMap<>();
        Set<String> direct = new HashSet<>();
        Set<String> visited = new HashSet<>();
        Deque<String> queue = new ArrayDeque<>();

        for (String imp : mainImports) {
            if (isStdlibImport(imp) || isLocal(imp, mainModulePath)) {
                continue;
            }
            ModVer mv = moduleOf(mods, imp);
            if (mv == null) {
                rs.complete = false;
                continue;
            }
            direct.add(mv.path);
            needed.put(mv.path, mv.version);
            queue.add(imp);
        }

        while (!queue.isEmpty()) {
            String imp = queue.poll();
            if (!visited.add(imp)) {
                continue;
            }
            if (isStdlibImport(imp) || isLocal(imp, mainModulePath)) {
                continue;
            }
            ModVer mv = moduleOf(mods, imp);
            if (mv == null) {
                // Maps to no build-list module: not part of the real build (e.g. a
                // //go:build ignore generator or an unselected platform). Diagnostic
                // only — not a missing dependency.
                rs.unresolved.add(imp);
                continue;
            }
            needed.put(mv.path, mv.version);

            List<String> deps = packageImports(src, mv.path, mv.version, imp);
            if (deps == null) {
                rs.complete = false;
                rs.missingDirs.add(imp);
                continue;
            }
            for (String dep : deps) {
                if (isStdlibImport(dep) || isLocal(dep, mainModulePath) || visited.contains(dep)) {
                    continue;
                }
                queue.add(dep);
            }
        }

        // For go < 1.17, omit indirect modules implied by another needed module's
        // go.mod (already pinned by the graph). go >= 1.17 records the full set.
        Set<String> implied = new HashSet<>();
        if (!separateIndirect) {
            for (ResolveResult.Edge e : res.graph()) {
                if (!e.fromPath.isEmpty() && !e.fromPath.equals(mainModulePath) && needed.containsKey(e.fromPath)) {
                    implied.add(e.toPath);
                }
            }
        }

        for (Map.Entry<String, String> en : needed.entrySet()) {
            String mod = en.getKey();
            if (direct.contains(mod)) {
                rs.direct.put(mod, en.getValue());
            } else if (!implied.contains(mod)) {
                rs.indirect.put(mod, en.getValue());
            }
        }
        return rs;
    }

    public static RequireSet tidyRequireSet(List<String> mainImports, String mainModulePath, String mainGoMod,
                                            ResolveResult res, ModSource src, boolean separateIndirect) {
        RequireSet base = neededModules(mainImports, mainModulePath, res, src, separateIndirect);
        if (!separateIndirect || !base.complete) {
            return base;
        }

        Map<String, String> loaded = new HashMap<>();
        List<ResolveResult.Module> mods = new ArrayList<>();
        for (ResolveResult.Module m : res.buildList()) {
            if (!m.main) {
                loaded.put(m.path, m.version);
                mods.add(m);
            }
        }

        // Current root set: everything NeededModules already requires.
        Map<String, String> roots = new HashMap<>();
        roots.putAll(base.direct);
        roots.putAll(base.indirect);

        ReqIndex idx = new ReqIndex(res, mainGoMod, src);

        // Walk the import graph FRONTIER BY FRONTIER (increasing import-stack
        // depth). At each frontier recompute the pruned selection under the roots
        // so far, then promote any frontier module the pruned graph under-selects.
        // This ordering pins a shallow module's deeper requirements before they are
        // examined, so they are not wrongly promoted. Test imports are deferred one
        // frontier deeper (go's separate `<pkg>.test` node).
        Set<String> queued = new HashSet<>();
        List<QItem> queue = new ArrayList<>();
        for (String imp : mainImports) {
            enq(imp, false, mainModulePath, queued, queue);
        }

        while (!queue.isEmpty()) {
            Map<String, String> sel = prunedSelectInMemory(roots, idx);
            List<QItem> frontier = queue;
            queue = new ArrayList<>();
            for (QItem it : frontier) {
                ModVer mv = moduleOf(mods, it.path);
                if (mv == null) {
                    continue;
                }
                PkgImports pi = packageImportsWithTests(src, mv.path, mv.version, it.path);
                if (pi != null) {
                    if (it.isTest) {
                        for (String d : pi.testImports) {
                            enq(d, false, mainModulePath, queued, queue);
                        }
                    } else {
                        for (String d : pi.imports) {
                            enq(d, false, mainModulePath, queued, queue);
                        }
                        enq(it.path, true, mainModulePath, queued, queue); // the test node, one frontier deeper
                    }
                }
                if (!roots.containsKey(mv.path)) {
                    String want = loaded.get(mv.path);
                    String have = sel.get(mv.path);
                    if (want != null && !want.isEmpty() && (have == null || GoSemver.compare(have, want) < 0)) {
                        roots.put(mv.path, want);
                    }
                }
            }
        }

        // Reclassify: direct stays direct; every other root is indirect.
        RequireSet out = new RequireSet();
        out.unresolved.addAll(base.unresolved);
        out.missingDirs.addAll(base.missingDirs);
        for (Map.Entry<String, String> e : roots.entrySet()) {
            if (base.direct.containsKey(e.getKey())) {
                out.direct.put(e.getKey(), e.getValue());
            } else {
                out.indirect.put(e.getKey(), e.getValue());
            }
        }
        return out;
    }

    // ---- helpers -----------------------------------------------------------

    private static List<ResolveResult.Module> nonMain(ResolveResult res) {
        List<ResolveResult.Module> mods = new ArrayList<>();
        for (ResolveResult.Module m : res.buildList()) {
            if (!m.main) {
                mods.add(m);
            }
        }
        return mods;
    }

    /** The longest build-list module path that is a prefix of importPath, or null. */
    private static @Nullable ModVer moduleOf(List<ResolveResult.Module> mods, String importPath) {
        String best = "";
        String bestVer = "";
        for (ResolveResult.Module m : mods) {
            if (importPath.equals(m.path) || importPath.startsWith(m.path + "/")) {
                if (m.path.length() > best.length()) {
                    best = m.path;
                    bestVer = m.version;
                }
            }
        }
        return best.isEmpty() ? null : new ModVer(best, bestVer);
    }

    private static boolean isLocal(String importPath, String mainModulePath) {
        return importPath.equals(mainModulePath) || importPath.startsWith(mainModulePath + "/");
    }

    /** A standard-library import: no dot in its first path segment. */
    static boolean isStdlibImport(String importPath) {
        int i = importPath.indexOf('/');
        String first = i >= 0 ? importPath.substring(0, i) : importPath;
        return !first.contains(".");
    }

    /** Non-test imports of a package, or null if its sources could not be read. */
    private static @Nullable List<String> packageImports(ModSource src, String mod, String version, String importPath) {
        Map<String, byte[]> files = src.packageGoFiles(mod, version, importPath);
        if (files == null) {
            return null;
        }
        Set<String> set = new HashSet<>();
        for (Map.Entry<String, byte[]> e : files.entrySet()) {
            if (e.getKey().endsWith("_test.go")) {
                continue;
            }
            for (String p : GoImports.parse(new String(e.getValue(), StandardCharsets.UTF_8))) {
                if (!p.isEmpty()) {
                    set.add(p);
                }
            }
        }
        return new ArrayList<>(set);
    }

    /** A package's ordinary and test imports, separately; null if unreadable. */
    private static @Nullable PkgImports packageImportsWithTests(ModSource src, String mod, String version, String importPath) {
        Map<String, byte[]> files = src.packageGoFiles(mod, version, importPath);
        if (files == null) {
            return null;
        }
        Set<String> imp = new HashSet<>();
        Set<String> test = new HashSet<>();
        for (Map.Entry<String, byte[]> e : files.entrySet()) {
            Set<String> target = e.getKey().endsWith("_test.go") ? test : imp;
            for (String p : GoImports.parse(new String(e.getValue(), StandardCharsets.UTF_8))) {
                if (!p.isEmpty()) {
                    target.add(p);
                }
            }
        }
        return new PkgImports(new ArrayList<>(imp), new ArrayList<>(test));
    }

    private static void enq(String path, boolean isTest, String mainModulePath, Set<String> queued, List<QItem> queue) {
        if (isStdlibImport(path) || isLocal(path, mainModulePath)) {
            return;
        }
        String k = isTest ? path + " t" : path;
        if (queued.add(k)) {
            queue.add(new QItem(path, isTest));
        }
    }

    private static final class QItem {
        final String path;
        final boolean isTest;

        QItem(String path, boolean isTest) {
            this.path = path;
            this.isTest = isTest;
        }
    }

    private static final class PkgImports {
        final List<String> imports;
        final List<String> testImports;

        PkgImports(List<String> imports, List<String> testImports) {
            this.imports = imports;
            this.testImports = testImports;
        }
    }

    /**
     * MVS-selected version of every module under the go1.17+ pruned graph rooted
     * at {@code roots}, reading requirements from {@code idx}. Mirrors {@link
     * Resolver} exactly: loaded modules' requires become nodes, but recursion
     * continues only through unpruned (go &lt; 1.17) modules.
     */
    static Map<String, String> prunedSelectInMemory(Map<String, String> roots, ReqIndex idx) {
        Map<String, String> present = new HashMap<>();
        Set<String> loadPath = new HashSet<>();
        Set<String> enqueued = new HashSet<>();
        Deque<ModVer> queue = new ArrayDeque<>();

        for (Map.Entry<String, String> e : roots.entrySet()) {
            ModVer m = new ModVer(e.getKey(), e.getValue());
            setNode(present, loadPath, enqueued, queue, m);
            loadPath.add(m.path);
            enqueue(enqueued, queue, m);
        }

        while (!queue.isEmpty()) {
            ModVer cur = queue.poll();
            if (!cur.version.equals(present.get(cur.path))) {
                continue;
            }
            ReqIndex.Reqs r = idx.requires(cur.path, cur.version);
            boolean unpruned = Resolver.goUnpruned(r.goVersion);
            for (ModVer req : r.reqs) {
                setNode(present, loadPath, enqueued, queue, req);
                if (unpruned) {
                    loadPath.add(req.path);
                    enqueue(enqueued, queue, req);
                }
            }
        }
        return present;
    }

    private static void enqueue(Set<String> enqueued, Deque<ModVer> queue, ModVer m) {
        if (enqueued.add(m.path + "@" + m.version)) {
            queue.add(m);
        }
    }

    private static void setNode(Map<String, String> present, Set<String> loadPath, Set<String> enqueued,
                                Deque<ModVer> queue, ModVer m) {
        String v = present.get(m.path);
        if (v == null || GoSemver.compare(m.version, v) > 0) {
            present.put(m.path, m.version);
            if (loadPath.contains(m.path)) {
                enqueue(enqueued, queue, m);
            }
        }
    }
}
